algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,246 @@
+
+                        s.spawn(move |_| {
+                            let now = std::time::Instant::now();
+                            let mut out = DeviceMemory::<Self>::zeros(ctx, n_gpu);
+                            P::scalar_mul_kernel(
+                                ctx,
+                                (n_gpu - 1) / cuda_group_size + 1, // grid
+                                cuda_group_size,     // block
+                                table.as_ptr(), exp.as_ptr(), out.as_mut_ptr(), n_gpu as isize
+                            )
+                            .expect("Kernel call failed");
+                            Self::batch_normalization(&mut out[..]);
+                            bases_gpu.clone_from_slice(&out.par_iter().map(|p| p.into_affine()).collect::<Vec<_>>()[..]);
+                            *time_gpu = now.elapsed().as_micros();
+                            println!("GPU {} finish", i);
+                        });
+                    }
+
+                    // Run on CPU
+                    s.spawn(|_| {
+                        let now = std::time::Instant::now();
+                        let exps_mut = &mut exps_h_ref.to_vec()[..];
+                        rayon::scope(|t| {
+                            for (b, s) in res_ref.chunks_mut(cpu_chunk_size).zip(exps_mut.chunks_mut(cpu_chunk_size)) {
+                                t.spawn(move |_| b[..].batch_scalar_mul_in_place(&mut s[..], 4));
+                            }
+                        });
+                        time_cpu = now.elapsed().as_micros();
+                        println!("CPU finish");
+                    });
+                });
+
+                // Update global microbenchmarking state
+                println!("old profile_data: {:?}", profile_data);
+                let cpu_throughput = n_cpu as f64 / time_cpu as f64;
+                let gpu_throughputs = n_gpus
+                    .iter()
+                    .zip(times_gpu.iter())
+                    .map(|(n_gpu, time_gpu)| {
+                        *n_gpu as f64 / *time_gpu as f64
+                })
+                .collect::<Vec<_>>();
+                let total_throughput = cpu_throughput + gpu_throughputs.iter().sum::<f64>();
+                let n_data_points = profile_data.1 as f64;
+                profile_data.1 += 1;
+                let new_proportions = gpu_throughputs.iter().map(|t| t / total_throughput);
+
+                if !profile_data.0.is_empty() {
+                    profile_data.0 = new_proportions.zip(profile_data.0.clone()).map(|(new, old)| {
+                        (new + n_data_points * old) / profile_data.1 as f64
+                    }).collect();
+                } else {
+                    profile_data.0 = new_proportions.collect();
+                }
+
+                // Update cached profiling data on disk
+                let now = std::time::Instant::now();
+                println!("writing data");
```

Comment:
This shouldn't appear in a normal run. Either change this to a debug print with logging or remove it.